lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
+                dim_coords = []
+                dim_name = self._mesh_dims[mesh][cube.location]
             else:
-                # No CF-netCDF coordinates describe this data dimension.
-                dim_name = "dim%d" % dim
-                if dim_name in self._existing_dim:
-                    # Increment name if conflicted with one already existing.
-                    if self._existing_dim[dim_name] != cube.shape[dim]:
+                # Get a name from the dim-coord (if any).
+                dim_coords = cube.coords(dimensions=dim, dim_coords=True)
+                if dim_coords:
+                    # Derive a dim name from a coord.
+                    coord = dim_coords[0]  # always have at least one
+
+                    # Add only dimensions that have not already been added.
+                    if coord in self._dim_names_and_coords.coords:
+                        # Return the dim_name associated with the existing
+                        # coordinate.
+                        dim_name = self._dim_names_and_coords.name(coord)
+                    else:
+                        # Determine a unique dimension name from the coord
+                        dim_name = self._get_coord_variable_name(cube, coord)
                         while (
                             dim_name in self._existing_dim
-                            and self._existing_dim[dim_name] != cube.shape[dim]
                             or dim_name in self._name_coord_map.names
                         ):
                             dim_name = self._increment_name(dim_name)
-                        # Update dictionary with new entry
-                        self._existing_dim[dim_name] = cube.shape[dim]
-                else:
-                    # Update dictionary with new entry
-                    self._existing_dim[dim_name] = cube.shape[dim]
 
-                dimension_names.append(dim_name)
-        return dimension_names
+                else:
+                    # No CF-netCDF coordinates describe this data dimension.
+                    # Make up a new, distinct dimension name
+                    dim_name = "dim%d" % dim
+                    if dim_name in self._existing_dim:
+                        # Increment name if conflicted with one already existing.
+                        if self._existing_dim[dim_name] != cube.shape[dim]:
+                            while (
+                                dim_name in self._existing_dim
+                                and self._existing_dim[dim_name]
+                                != cube.shape[dim]
+                                or dim_name in self._name_coord_map.names
+                            ):
+                                dim_name = self._increment_name(dim_name)
+
+            # Record the dimension.
+            record_dimension(dim_name, cube.shape[dim], dim_coords)
+
+        cube_dimensions = dimension_names.copy()
+
+        return cube_dimensions, mesh_dimensions
```

Comment:
Minor point, but this could just be:
```python
return dimension_names, mesh_dimensions
```
there's no need for the `copy` or `cube_dimensions`... but if you think it's clearer as is, I'm okay with that too

Reply:
:+1:
Since our talk yesterday, I now have a plan for simplifying this